src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -0,0 +1,1341 @@
+    static object P => new object();
+    static object? Q => null;
+    static void G()
+    {
+        F(P).ToString();
+        F(Q).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(Q).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(Q)").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void TypeInference_FieldAccess()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static object F1 = new object();
+    static object? F2 = null;
+    static void G()
+    {
+        F(F1).ToString();
+        F(F2).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(F2).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(F2)").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void TypeInference_Literal()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G()
+    {
+        F(0).ToString();
+        F('A').ToString();
+        F(""B"").ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TypeInference_Default()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G()
+    {
+        F(default(object)).ToString();
+        F(default(int)).ToString();
+        F(default(string)).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (6,9): warning CS8602: Possible dereference of a null reference.
+                //         F(default(object)).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(default(object))").WithLocation(6, 9),
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         F(default(string)).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(default(string))").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void TypeInference_Tuple()
+        {
+            var source =
+@"class C
+{
+    static (T, U) F<T, U>((T, U) t) => t;
+    static void G(string x, string? y)
+    {
+        var t = (x, y);
+        F(t).Item1.ToString();
+        F(t).Item2.ToString();
```

Comment:
Could you add two variants: one using explicit names `var t = (a: x, b: y); F(t).a.ToString();` and one using inferred names `var t = (x, y); F(t).x.ToString();` ?
I'd expect warnings any time we use the second field.

It may also be worthwhile to also test a long tuple.

Reply:
Added test for explicit and inferred names.